// ngx-scanner-qrcode/esm5/lib/ngx-scanner-qrcode.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { AsyncSubject } from 'rxjs';
import { AS_COMPLETE, FILES_TO_SCAN } from './ngx-scanner-qrcode.helper';
import * as i0 from "@angular/core";
var NgxScannerQrcodeService = /** @class */ (function () {
    function NgxScannerQrcodeService() {
    }
    /**
     * Load files
     * @param files
     * @return AsyncSubject
     */
    /**
     * Load files
     * @param {?=} files
     * @return {?} AsyncSubject
     */
    NgxScannerQrcodeService.prototype.loadFiles = /**
     * Load files
     * @param {?=} files
     * @return {?} AsyncSubject
     */
    function (files) {
        var _this = this;
        if (files === void 0) { files = []; }
        /** @type {?} */
        var as = new AsyncSubject();
        Promise.all(Object.assign([], files).map((/**
         * @param {?} m
         * @return {?}
         */
        function (m) { return _this.readAsDataURL(m); }))).then((/**
         * @param {?} img
         * @return {?}
         */
        function (img) { return AS_COMPLETE(as, img); })).catch((/**
         * @param {?} error
         * @return {?}
         */
        function (error) { return AS_COMPLETE(as, null, error); }));
        return as;
    };
    /**
     * loadFilesToScan
     * @param files
     * @param config
     * @returns
     */
    /**
     * loadFilesToScan
     * @param {?=} files
     * @param {?=} config
     * @return {?}
     */
    NgxScannerQrcodeService.prototype.loadFilesToScan = /**
     * loadFilesToScan
     * @param {?=} files
     * @param {?=} config
     * @return {?}
     */
    function (files, config) {
        if (files === void 0) { files = []; }
        return FILES_TO_SCAN(files, config);
    };
    /**
     * readAsDataURL
     * @param file
     * @return Promise
     */
    /**
     * readAsDataURL
     * @private
     * @param {?} file
     * @return {?} Promise
     */
    NgxScannerQrcodeService.prototype.readAsDataURL = /**
     * readAsDataURL
     * @private
     * @param {?} file
     * @return {?} Promise
     */
    function (file) {
        /** drawImage **/
        return new Promise((/**
         * @param {?} resolve
         * @param {?} reject
         * @return {?}
         */
        function (resolve, reject) {
            /** @type {?} */
            var fileReader = new FileReader();
            fileReader.onload = (/**
             * @return {?}
             */
            function () {
                /** @type {?} */
                var objectFile = {
                    name: file.name,
                    file: file,
                    url: URL.createObjectURL(file)
                };
                resolve(objectFile);
            });
            fileReader.onerror = (/**
             * @param {?} error
             * @return {?}
             */
            function (error) { return reject(error); });
            fileReader.readAsDataURL(file);
        }));
    };
    NgxScannerQrcodeService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */ NgxScannerQrcodeService.ngInjectableDef = i0.defineInjectable({ factory: function NgxScannerQrcodeService_Factory() { return new NgxScannerQrcodeService(); }, token: NgxScannerQrcodeService, providedIn: "root" });
    return NgxScannerQrcodeService;
}());
export { NgxScannerQrcodeService };
